File: catkin_ws/src/omnibot/src/omnibot/pid.py
class Pid:
    i_err = 0
    d_err = 0
    last_err = 0
    pos = 0
    sp = 0
    
    def __init__(self, kp=0, ki=0, kd=0) -> None:
        self.kp = kp
        self.ki = ki
        self.kd = kd

    def pid(self) -> float:
        err = self.sp - self.pos
        self.d_err = err - self.last_err
        self.last_err = err
        self.i_err = self.i_err + err
        result_pid = (self.kp * err) + (self.ki * self.i_err) + (self.kd * self.d_err)
        return result_pid

    def reset_err(self):
        self.i_err = 0
        self.d_err = 0
        self.last_err = 0
# p = Pid()
# p.kp = 0.1
# p.sp = 10
# while True:
#     p.pos += p.pid()
#     print(p.pos)
#     sleep(0.1)

File: ESP32/lib/pid/src/pid.cpp
#include "pid.h"
/*
    Mengatur nilai Kp, Ki, dan Kd
*/
Pid::Pid(float kp, float ki, float kd)
{
    this->kp = kp;
    this->ki = ki;
    this->kd = kd;
}

/*
    Menghitung nilai PWM berdasarkan error yang ada
    (error titik pengesetan, error integral, dan error derivatif)
*/
float Pid::compute(float now)
{
    if (i_err > windup)
        i_err = 0;
    float err = setpoint - now;
    d_err = err - last_err;
    last_err = err;
    i_err = i_err + err;
    float result = (kp * err) + (ki * i_err) + (kd * d_err);
    if (result >= 0)
    {
        return result > limit ? limit : result;
    }
    else {
        return result < - limit ? - limit : result;
    }
}
/*
    Mengatur ulang semua perhitungan error
*/
void Pid::reset()
{
    d_err = 0;
    i_err = 0;
    last_err = 0;
    setpoint = 0;
}
/*
    Mengubah nilai Kp, Ki, Kd
*/
void Pid::setPid(float kp, float ki, float kd)
{
    this->kp = kp;
    this->ki = ki;
    this->kd = kd;
}